// factory.ts/lib/spec/pipeline.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const Pipe = require("../src/pipeline");
const Factory = require("../src/async");
describe("pipelines", () => {
    const childFactory = Factory.makeFactory({
        name: "Kid",
        grade: 1
    });
    const parentFactory = Factory.makeFactoryWithRequired({
        name: "Parent",
        birthday: Factory.each(i => Promise.resolve(new Date(`2017/05/${i}`))),
        children: Factory.each(() => [])
    }, { startingSequenceNumber: 1 });
    const grandpaFactory = parentFactory.transform(parent => {
        return Object.assign(Object.assign({}, parent), { spoils: true });
    });
    it("builds data in steps", () => __awaiter(void 0, void 0, void 0, function* () {
        const p = Pipe.Pipeline.start()
            .addValues({ hello: "kitty", hola: "espanol" })
            .addValues(() => Promise.resolve({ byebye: "birdie" }))
            .addValues(v => ({
            corner: `${v.hello} corner`,
            golf: v.byebye
        }))
            .addFactory(childFactory, "kiddo", { grade: 2 })
            .addFactory(parentFactory, "dad", v => Promise.resolve({
            name: "Dad",
            children: [v.kiddo],
            spouse: null
        }))
            .addTxFactory(grandpaFactory, "gramps", v => ({
            name: "Gramps",
            children: [v.dad],
            spouse: null
        }));
        const data = yield p;
        expect(data.hello).toEqual("kitty");
        expect(data.hola).toEqual("espanol");
        expect(data.byebye).toEqual("birdie");
        expect(data.corner).toEqual("kitty corner");
        expect(data.golf).toEqual("birdie");
        expect(data.kiddo.grade).toEqual(2);
        expect(data.kiddo.name).toEqual("Kid");
        expect(data.dad.name).toEqual("Dad");
        expect(data.dad.birthday.getTime()).toEqual(new Date("2017/05/01").getTime());
        expect(data.dad.children.length).toEqual(1);
        expect(data.dad.children[0]).toEqual(data.kiddo);
        expect(data.dad.spouse).toBeNull();
        expect(data.gramps.name).toEqual("Gramps");
        expect(data.gramps.spoils).toEqual(true);
        expect(data.gramps.children.length).toEqual(1);
        expect(data.gramps.children[0]).toEqual(data.dad);
        expect(data.gramps.spouse).toBeNull();
    }));
});
